File: backend-go/internal/services/translate.go
package services

import (
	"context"
	"encoding/json"
	"net/http"
	"net/url"
	"strings"

	"github.com/centraluniversity/researcher/internal/models"
)

func Translate(ctx context.Context, llm LLM, paper models.PaperOut, text, target string) (string, *string) {
	text = strings.TrimSpace(text)
	if text == "" {
		return "", nil
	}
	if llm.Config.LLMAPIKey != "" {
		reply := llm.Explain(ctx, paper, text, "Translate the fragment into "+target+". Return only the translation, no quotes or commentary.")
		l := strings.ToLower(reply)
		if reply != "" && !strings.HasPrefix(l, "ошибка") && !strings.Contains(l, "не настроен") {
			return reply, nil
		}
	}
	u := "https://api.mymemory.translated.net/get?q=" + url.QueryEscape(truncate(text, 450)) + "&langpair=" + url.QueryEscape("autodetect|"+target)
	req, _ := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "Не удалось перевести: " + err.Error(), nil
	}
	defer resp.Body.Close()
	var data struct {
		ResponseData struct {
			TranslatedText string `json:"translatedText"`
		} `json:"responseData"`
		Matches []struct {
			Source string `json:"source"`
		} `json:"matches"`
	}
	if err = json.NewDecoder(resp.Body).Decode(&data); err != nil || strings.TrimSpace(data.ResponseData.TranslatedText) == "" {
		return "Не удалось перевести: empty translation from MyMemory", nil
	}
	var source *string
	if len(data.Matches) > 0 && data.Matches[0].Source != "" {
		source = &data.Matches[0].Source
	}
	return strings.TrimSpace(data.ResponseData.TranslatedText), source
}
func truncate(s string, n int) string {
	r := []rune(s)
	if len(r) > n {
		return string(r[:n])
	}
	return s
}

File: backend/app/services/translate.py
from __future__ import annotations

import logging

import httpx

from app.core.config import settings
from app.services import llm as llm_service
from app.models import Paper

logger = logging.getLogger(__name__)


def _translate_mymemory(text: str, target_lang: str) -> tuple[str, str | None]:
    """Бесплатный fallback без гео-блоков LLM (MyMemory)."""
    langpair = f"autodetect|{target_lang}"
    with httpx.Client(timeout=20.0) as client:
        response = client.get(
            "https://api.mymemory.translated.net/get",
            params={"q": text[:450], "langpair": langpair},
        )
        response.raise_for_status()
        data = response.json()
    translated = (data.get("responseData") or {}).get("translatedText") or ""
    detected = None
    matches = data.get("matches") or []
    if matches and isinstance(matches[0], dict):
        detected = matches[0].get("source")
    if not translated.strip():
        raise RuntimeError("Empty translation from MyMemory")
    return translated.strip(), detected


def translate_text(paper: Paper, text: str, target_lang: str = "ru") -> tuple[str, str | None]:
    """
    Перевод выделения.
    Сначала пробуем LLM (если ключ есть), иначе MyMemory.
    """
    cleaned = text.strip()
    if not cleaned:
        return "", None

    if settings.llm_api_key:
        reply = llm_service.explain_fragment(
            paper,
            cleaned,
            question=(
                f"Translate the fragment into {target_lang}. "
                "Return only the translation, no quotes or commentary."
            ),
        )
        # если LLM вернул ошибку провайдера — уйдём в fallback
        lowered = reply.lower()
        if reply and not lowered.startswith("ошибка") and "не настроен" not in lowered:
            return reply.strip(), None

    try:
        return _translate_mymemory(cleaned, target_lang)
    except Exception as exc:  # noqa: BLE001
        logger.exception("Translation failed")
        return f"Не удалось перевести: {exc}", None
